### PR TITLE
Replace `as` casts in llvm.x86.addcarry.64 implementation

### PR DESCRIPTION
**Pull request by dtolnay**
Recommended in https://github.com/rust-lang/miri/pull/2090#pullrequestreview-958491177.

### CODE REVIEW

**Comment by RalfJung** [comment]
Fancy. :D  Thanks a lot!

Can you move the new file into the (existing) `helpers` module, i.e., create a subdirectory with that name and put it there?

File: src/helpers/convert.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+mod implementations {
+    pub(crate) trait NarrowerThan<T> {
+        fn truncate_from(wide: T) -> Self;
+    }
+
+    macro_rules! impl_narrower_than {
+        ($(NarrowerThan<{$($ty:ty),*}> for $self:ty)*) => {
+            $($(
+                impl NarrowerThan<$ty> for $self {
+                    fn truncate_from(wide: $ty) -> Self {
```
Comment (RalfJung) [inline src/helpers/convert.rs:35]: Do you think these should be `inline(always)` just to be sure?

**Comment by RalfJung** [comment]
r=me with or without the `inline`
@bors delegate+

File: src/helpers/convert.rs (hunk 1 above)
Comment (dtolnay) [reply src/helpers/convert.rs:35]: Even if this code were in my most performance-sensitive crates, I would not expect to need `inline(always)` here. In the standard library even `From<T> for T` and `From<u8> for u16` do not have `inline(always)`.

**Pull request merged**